rustdoc: Consolidate static-file replacement mechanism

jsha opened the pull request:
There were a few places in rustdoc where we would take static JS or CSS and rewrite it at doc generation time to insert values. This consolidates all the CSS instances into one CSS file and replaces the JS examples with data- attributes on the rustdoc-vars div.

Demo https://rustdoc.crud.net/jsha/static-file-replace/test_docs/

r? @GuillaumeGomez

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4270415944?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b424fd8c641801256112472b3e8a9104e9b03630/checks/4270415944/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
ES-Check: there were no ES version matching errors!  🎉
+ eslint ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/source-script.js ../src/librustdoc/html/static/js/storage.js

/checkout/src/librustdoc/html/static/js/main.js
  142:5  error  'availableThemes' is not defined  no-undef
/checkout/src/librustdoc/html/static/js/storage.js
/checkout/src/librustdoc/html/static/js/storage.js
  2:11  error  'resourcesSuffix' is defined but never used  no-unused-vars

✖ 2 problems (2 errors, 0 warnings)
```

</details>

jyn514 commented on src/librustdoc/html/render/write_shared.rs:
```diff
@@ -247,64 +210,83 @@ pub(super) fn write_shared(
+        options,
+    )?;
+
     // FIXME: this should probably not be a toolchain file since it depends on `--theme`.
```
You can remove this FIXME now I think :)

jyn514 commented on src/librustdoc/html/render/write_shared.rs:
```diff
@@ -247,64 +210,83 @@ pub(super) fn write_shared(
     // FIXME: this should probably not be a toolchain file since it depends on `--theme`.
     // But it seems a shame to copy it over and over when it's almost always the same.
     // Maybe we can change the representation to move this out of main.js?
     write_minify(
```
Hmm, I think this should be using `cx.write_minify(InvocationSpecific)`, since it can vary from crate to crate. Unfortunately we don't have a good way of testing this before it hits nightly right now.

jsha replied:
The only thing that is invocation specific is the list of themes, right? What if I put those in a `data-` attribute in the HTML instead? The list is quite short, so duplication is not a big concern.

jyn514 replied:
That would be great! Thanks :)

GuillaumeGomez commented on src/librustdoc/html/static/js/storage.js:
```diff
@@ -1,5 +1,8 @@
 // From rust:
 /* global resourcesSuffix */
+var resourcesSuffix = /* AUTOREPLACE: */"RESOURCE_SUFFIX";
```
I'm a bit worried about this: if we generate docs with two different rustdoc version and/or different suffix, it means the older version will use newer files, which might break the display.

jsha replied:
storage is stored as a versioned file (e.g. storage1.58.0.js), so it would load the other versioned files that match it.

GuillaumeGomez replied:
That is, if you use suffix, otherwise you'll simply overwrite.

jyn514 replied:
I don't think it's possible for rustdoc alone to prevent conflicting versions with our current setup. If you think that should be a goal, maybe open an issue or MCP but I don't think it makes sense to hold specific parts of rustdoc to a higher standard than others.

jsha replied:
The more I think about it, the more I think it would be cleaner to put each of these three variables in a `data-` attribute in the HTML, like I proposed in https://github.com/rust-lang/rust/pull/91062#discussion_r753640831. The big advantage there is that when you're editing JS you can copy it directly into place without waiting to rebuild rustdoc. I've run into this a number of times just recently.

GuillaumeGomez replied:
> I don't think it's possible for rustdoc alone to prevent conflicting versions with our current setup. If you think that should be a goal, maybe open an issue or MCP but I don't think it makes sense to hold specific parts of rustdoc to a higher standard than others.

It is working currently because the links are set directly in the HTML, not in the JS which is regenerated. This is my biggest issue with this change, moving "hard-coded" HTML content into JS.

@jsha Yes, I think that could work. The added size would be negligible as well.

jsha commented:
Alright, I switched this up so that main.js isn't rewritten at all - instead, the variables we want are set in the rustdoc-vars div.

One big change along the way: I revamped how stylesheets links are rendered into the HTML (https://github.com/rust-lang/rust/pull/91062/commits/c3309634449ccb021f5d78bfaeeb6211e093c2ac). I think this cleans things up a lot. It also made it much nicer to be able to generate a list of themes in layout.rs.

Also I changed storage.js to not rely on resource_suffix, because rustdoc-vars aren't yet available at the time storage.js runs.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4308167076?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/7a5cd2be84623011dd7d4d77e510a39d7c8bf657/checks/4308167076/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [rustdoc] rustdoc/static-root-path.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/static-root-path" "/checkout/src/test/rustdoc/static-root-path.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
6: @matches check failed
 `PATTERN` did not match
 // @matches - '"\.\./search-index\.js"'
Encountered 1 errors

------------------------------------------

---
test result: FAILED. 479 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out; finished in 55.60s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:17:03
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4308532786?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/276ea817a2d2711ec6055f47562e9285fb808672/checks/4308532786/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between de4b242e1e2143f549f25ac5a8f7de9d902ef3b4 and eddbe84c38e27021f2b4fc5fce9e1b683ca60789
Rustdoc was updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
......... (50/53)
...        (53/53)


/checkout/src/test/rustdoc-gui/toggle-click-deadspace.goml toggle-click-deadspace... FAILED
[ERROR] (line 8) Error: Evaluation failed: assert didn't fail: for command `assert-attribute-false: (".impl-items .rustdoc-toggle", {"open": ""})`



command did not execute successfully: "/node-v14.4.0-linux-x64/bin/node" "/checkout/src/tools/rustdoc-gui/tester.js" "--jobs" "16" "--doc-folder" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/doc" "--tests-folder" "/checkout/src/test/rustdoc-gui"


Build completed unsuccessfully in 0:00:18
```

</details>

jyn514 reviewed (approved):
LGTM but I want to give Guillaume a chance to review. Thanks for working on this :)

GuillaumeGomez commented:
It looks good to me as well, thanks!  You'll need to rebase to make the CI happy. ;)

r=jyn514,GuillaumeGomez once done.

jsha commented:
@bors r=jyn514,GuillaumeGomez

bors commented:
:pushpin: Commit d9afca58081e09c7dd9fa5c874810e05dce18d49 has been approved by `jyn514,GuillaumeGomez`

<!-- @bors r=jyn514,GuillaumeGomez d9afca58081e09c7dd9fa5c874810e05dce18d49 -->
<!-- homu: {"type":"Approved","sha":"d9afca58081e09c7dd9fa5c874810e05dce18d49","approver":"jyn514,GuillaumeGomez"} -->

Folyd commented:
Ohhh, this PR broke my extension https://github.com/huhu/rust-search-extension/issues/134. 😅 
**Rust search extension** relies on the `search-index.js`. I hope we can keep the `search-index.js` loading steadily. Would you mind CC me if the `search-index.js` loading method has changed? Thanks. ❤️  @jsha @GuillaumeGomez

jyn514 commented:
@Folyd you can set up an automatic ping in the highfive bot if you like - here's an example: https://github.com/rust-lang/highfive/blob/6e2c21639aaeafaeae423b244d353247c507d46a/highfive/configs/rust-lang/rust.json#L134
